### ISSUE TITLE
Style/HashSlice breaks element order

### ISSUE THREAD

**Issue by forthrin** [rubocop/rubocop#14659]
## Steps to reproduce the problem

```ruby
hash = { a: 1, b: 2, c: 3 }
want = %i[b a]

hash.select { |key, _| want.include?(key) }.then { pp it }
hash.slice(*want).then { pp it }

# {a: 1, b: 2} # order of hash
# {b: 2, a: 1} # order of want
```
## Expected behavior

Do nothing. Or is there a terser way to get the elements in `hash` order, do tell!

## Actual behavior

```Use `slice(*want)` instead. [[Correctable] Style/HashSlice]```

## RuboCop version

```
System Version: macOS 26.0.1 (25A362)
Model Identifier: MacBookAir10,1
ruby      2025-08-01 30a20bc16
rubocop   2025-10-18 f7f653f50
```

**Comment by koic** [comment]
This is a safety issue. The result of `Hash#slice` depends on the order of the arguments passed to `slice`, such as `want = %i[b a]`. In other words, changing this case to `want = %i[a b]` or using `hash.slice(*want.reverse).then { pp it }` yields the expected order. However, because this depends on the arguments to `slice`, there is no universal solution.

`Style/HashSlice` is already marked as unsafe. Also, an ordered hash is not always required, and whether this becomes an issue depends on the use case. So, it should be better to supplement the documentation with notes about safety regarding result order, rather than changing its behavior.

**Comment by forthrin** [comment]
@koic: Updated OP with more arbitrary values to clarify the discrepancy. Not sure I follow the reasoning for upholding the suggested rewrite when the outcome is different and the original code may be justified some times.

**Comment by koic** [comment]
I see. As mentioned earlier, Hash object is not always intended to be used as an ordered hash. On the other hand, I understand the desire to apply the cop as safely as possible. With that in mind, it might be worth considering providing an option for cases like the one shown in the example.

**Comment by forthrin** [comment]
Sounds better. Give due contemplation before settling. Albeit circuitous, Ruby hashes ultimately appear deterministic.

https://ruby-doc.org/3.4.1/Hash.html#class-Hash-label-Entry+Order
https://stackoverflow.com/questions/31418673/is-order-of-a-ruby-hash-literal-guaranteed

**Comment by forthrin** [comment]
Ran into this one again today. Thing is that `hash` may be an Array for all we know, and the `select` will still work, because Ruby treats a Hash and an Array with two subelements gracefully equal a lot of the time.

However, Rubocop's suggestion to use `slice` will break the code if it is in fact an Array. So that's another reason why this cop should surrender his badge and take up stamp collecting and topiary or something.

```ruby
want = %i[h f t]
hash1 = { f: 1, g: 9, t: 8, h: 4 }
hash2 = [[:f, 1], [:g, 9], [:t, 8], [:h, 4]]

[hash1, hash2].each do |hash|
  puts "\033[92m#{hash}\033[m"
  begin
    hash.slice(*want).then { pp it }
  rescue ArgumentError
    puts "\033[101m#{hash} # POLICE BRUTALITY!!!!\033[m"
  end
  hash.select { |key, _| want.include?(key) }.then { pp it }
  want.filter_map { |w| hash.find { it.first == w } }.then { pp it }
  puts '-' * 70
end
```
